src/java.base/share/classes/java/util/Collections.java

```diff
@@ -1443,7 +1467,13 @@ private Object writeReplace() {
      * @param  m the map for which an unmodifiable view is to be returned.
      * @return an unmodifiable view of the specified map.
      */
+    @SuppressWarnings("unchecked")
     public static <K,V> Map<K,V> unmodifiableMap(Map<? extends K, ? extends V> m) {
+        if(m.getClass() == UnmodifiableMap.class ||
+           m.getClass() == ImmutableCollections.Map1.class ||
```

Comment:
(I'm not a reviewer.)

I think this causes a change in behavior to this silly program.

```java
var map1 = Map.of(1, 2);
var map2 = Collections.unmodifiableMap(map1);

try {
  System.out.println("map1 returned " + map1.entrySet().contains(null));
} catch (NullPointerException e) {
  System.out.println("map1 threw");
}

try {
  System.out.println("map2 returned " + map2.entrySet().contains(null));
} catch (NullPointerException e) {
  System.out.println("map2 threw");
}
```

With JDK 15 the output is:
> map1 threw
> map2 returned false

With this change I think the output will be:
> map1 threw
> map2 threw

It seems unlikely that anyone will be bit by this, but it is a change to behavior and it wasn't called out in the Jira issue, so I felt it was worth mentioning.

I think it is just this one specific case that changes -- only `Map1`, and only `entrySet().contains(null)`.  Other sub-collections like `keySet()` and `values()` and `subList(...)` already throw on `contains(null)` for both the `ImmutableCollections.*` implementation and the `Collections.umodifiable*` wrapper.  `MapN`'s `entrySet().contains(null)` already returns `false` for both.

Reply:
This sounds like an inconsistency between `Map1` and `MapN` that should perhaps be considered a bug that needs fixing. /ping @stuart-marks

Reply:
2 remarks:
1. MapN's entry set seems to extend abstract set, whose `contains` is null-friendly like https://github.com/openjdk/jdk/blob/cb84539d56209a6687c4ec71a61fdbe6f06a46ea/src/java.base/share/classes/java/util/AbstractCollection.java#L104
2. The problem of unmodifiable map's entry set not always delegating everything to the backing entry set still exists. https://github.com/openjdk/jdk/blob/cb84539d56209a6687c4ec71a61fdbe6f06a46ea/src/java.base/share/classes/java/util/Collections.java#L1724 This will bypass the underlying logic when the argument is `null` or not an entry.

Reply:
This raises some interesting issues and makes me wonder if we should allow a single-wrap of the `ImmutableCollections` classes for now to make this less onerous.

Reply:
Yes -- I think in response to this it makes more sense to pull the `ImmutableCollections` classes out for now and only focus on the wrapping of the classes within `Collections` so we aren't blocked by studying and rectifying these inconsistencies.

Reply:
Maybe it is not correct or UnmodifiableEntrySet::contains to short circuit?  What if the implementation was changed to:

`public boolean contains(Object o) {
            if (!(o instanceof Map.Entry))
                return c.contains(o); //false, NPE, or CCE
            return c.contains(
                new UnmodifiableEntry<>((Map.Entry<?,?>) o));
}`

Reply:
This however changes the behavior of unmodifiable maps compared to before; i.e. before for other entry sets, they could not throw exception if the object passed was not map entry; now they can.

Reply:
Yes, it seems reasonable to decouple the `ImmutableCollections` issues from this change.